src/pynwb/validate.py

```diff
@@ -1,137 +1,223 @@
-                                help="Use the cached namespace (default).")
-    feature_parser.add_argument('--no-cached-namespace', dest="cached_namespace", action='store_false',
-                                help="Don't use the cached namespace.")
-    parser.set_defaults(cached_namespace=True)
+    # Determine which namespaces are the most specific (i.e. extensions) and validate against those
+    candidate_namespaces = set(namespace_dependencies.keys())
+    for namespace_dependency in namespace_dependencies:
+        candidate_namespaces -= namespace_dependencies[namespace_dependency].keys()
 
-    args = parser.parse_args()
-    ret = 0
-
-    # TODO Validation against a specific namespace file is currently broken. See pynwb#1396
-    # if args.nspath:
-    #     if not os.path.isfile(args.nspath):
-    #         print("The namespace file {} is not a valid file.".format(args.nspath), file=sys.stderr)
-    #         sys.exit(1)
-    #
-    #     if args.cached_namespace:
-    #         print("Turning off validation against cached namespace information "
-    #               "as --nspath was passed.", file=sys.stderr)
-    #         args.cached_namespace = False
-
-    for path in args.paths:
-
-        if not os.path.isfile(path):
-            print("The file {} does not exist.".format(path), file=sys.stderr)
-            ret = 1
-            continue
-
-        if args.cached_namespace:
-            catalog = NamespaceCatalog(NWBGroupSpec, NWBDatasetSpec, NWBNamespace)
-            ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
-            s = set(ns_deps.keys())       # determine which namespaces are the most
-            for k in ns_deps:             # specific (i.e. extensions) and validate
-                s -= ns_deps[k].keys()    # against those
-            # TODO remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
-            if 'hdmf-experimental' in s:
-                s.remove('hdmf-experimental')  # remove validation of hdmf-experimental for now
-            namespaces = list(sorted(s))
-            if len(namespaces) > 0:
-                tm = TypeMap(catalog)
-                manager = BuildManager(tm)
-                specloc = "cached namespace information"
+    # TODO: remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
+    candidate_namespaces.discard("hdmf-experimental")  # remove validation of hdmf-experimental for now
+    cached_namespaces = sorted(candidate_namespaces)
+
+    if len(cached_namespaces) > 0:
+        type_map = TypeMap(namespaces=catalog)
+        manager = BuildManager(type_map=type_map)
+    else:
+        manager = None
+
+    return cached_namespaces, manager, namespace_dependencies
+
+
+@docval(
+    {
+        "name": "io",
+        "type": (HDMFIO, type(None)),  # want to do Optional[HDMFIO] but docval complains?
```

Comment:
```suggestion
        "type": HDMFIO,
```
Having this plus a `"default"` value will make the argument optional. Does that accomplish what you are intending?

Reply:
Yep